ImageView Parameters on GPU

danieldresser-ie opened the pull request:
This has already been previewed and discussed a fair bit, so hopefully it's about ready to go.

It is on by default, and somewhat invasive, so it would be good to get some careful review and testing.

It does help performance in general some, and it definitely makes adjusting exposure while looking at an image a whole lot nicer.

The quality of the result generally seems very good when comparing with turning on the CPU fallback, though it sometimes requires tuning the allocationvars in your OCIO config.

The only thing I'm currently aware of being wrong is that the DisplayTransform dropdown isn't large enough to display all the text it needs to ( along with the way I'm hacking up the menu button text on the PresetsPlugValueWidget being extremely ugly ), but Tom suggested I shouldn't worry about the UI side at the moment.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -45,6 +45,59 @@
+	def __toggleGpu( self, v ) :
+		self.__gpuPlug.setValue( not self.__gpuPlug.getValue() )
+
+	## \todo : Duplicating this code is very ugly.  Need a better way to override the menu
+	##         button text
+	def _updateFromPlug( self ) :
+
+		self._PresetsPlugValueWidget__menuButton.setEnabled( self._editable() )
+
+		text = ""
+		if self.getPlug() is not None :
+			with self.getContext() :
+				presetName = Gaffer.NodeAlgo.currentPreset( self.getPlug() )
+
+			text = presetName or "Invalid"
+
+		if self.__gpuPlug.getValue():
+			text += " (GPU)"
+		else:
+			text += " (slow)"
+
+		self._PresetsPlugValueWidget__menuButton.setText( text )
+		self._PresetsPlugValueWidget__customValuePlugWidget.setVisible( False )
```
Can't we just call `PresetsPlugValueWidget._updateFromPlug( self )` and then call `menuButton.setText( menuButton.getText() + "our suffix" )`?

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -45,6 +45,59 @@
+		if self.__gpuPlug.getValue():
+			text += " (GPU)"
+		else:
+			text += " (slow)"
```
CPU?

danieldresser-ie replied:
I don't much like how few pixel of difference there are between "GPU" and "CPU", especially since I was thinking it might be nice to make this text smaller, but I've gone with CPU for now, see what you think of it it actually runs on your system.

themissingcow replied:
Generally speaking, we always try to be consistent with pairings.  So grammatically speaking, it's the same type of word. In that, if one is describing 'what it runs on', then so should the other state. If you were to  go with `slow` - is an adjective describing the result, then GPU should change to something like `fast`. Though I think then it might be better to use `accurate` over `slow`, as it better helps the user understand why they'd choose one or the other.

It seems people are already using standard parenthesis in production as part of their names eg, from CS: `Default (show display without primary)`. We don't want to overload this title with info, and we want GPU to be the default right? As such, how about:

 - Use square brackets `[CPU]`
 - Only append this if on the CPU path, and leave it as-is for GPU. 

Then the common presentation is less cluttered, and its more apparent when you're on something custom? So it'd be:

`Default` or `Default [CPU]`

themissingcow replied:
People seem to want to know which is which. 

How about this. Keep the dropdown exactly as it was, add a small LUT cube icon next to it to the left. That presents a menu with the processing options in.  This fits the pattern we have in the SceneView (icons = menu) and ImageView ( icons = clickable ), it gets rid of the original concern that were abusing a combo box, and allows us mutually exclusive items (eg `GPU` and `CPU`), and to adapt the icon to reflect CPU/GPU state.

It also means that peoples customisations of the current widget (to display the actual profile in `Default`) will just keep on working too.

danieldresser-ie commented on include/GafferImageUI/ImageGadget.h:
```diff
@@ -118,6 +125,21 @@ class GAFFERIMAGEUI_API ImageGadget : public GafferUI::Gadget
+		float getGamma();
+
+		void setDisplayTransform( GafferImage::ImageProcessorPtr displayTransform );
+		GafferImage::ImageProcessorPtr getDisplayTransform();
```
Did you mean you wanted both a const and non-const version, or is just a const one enough?

Also, it might be worth talking about the ways in which this is kinda sketchy?  For the GPU path, we just call transform() on the node, and it could actually be const.  But for the CPU path, we actually are reconnecting the inputs of the provided displayTransform node - this API doesn't make it overly obvious that passing in a displayTransform that is reused would be very bad.  I guess it's fine, because it's mostly used from ImageView which keeps it's own internal list of display transforms which isn't used for anything else?

johnhaddon replied:
I think I was suggesting a const version in addition to the non-const version, but I see you've gone with const-only now, and I think that was the right choice. We don't want external code to modify the processor after setting it, and providing only const access gives us a little more protection in that respect.

danieldresser-ie commented:
I think I've addressed all John's specific comments ( except for the possibility that R16F isn't supported on Mac ), and I've added the dependency on OCIO in the SConstruct which appears to have gotten it to build on mac in the CI ( did John hack this in for a test build previously? ).

Hopefully the glGenTextures fix gets it working on Mac.

But I still need to figure out why I'm getting GadgetWidgets not being cleaned up in all the tests.

The commit history is kinda messed up at the moment - I was trying to do fixes as separate commits to avoid needing to force push before the commits get reviewed, but then a conflict in Changes.md forced me to force push anyway.

themissingcow commented:
> Thanks Daniel - I'm very much looking forward to testing this! Unfortunately it crashes for me at present, and the CI seems to be crashing too. I'm afraid I haven't been able to get to the bottom of that, but I did comment one potential cause inline. In my absence, could you coordinate with @themissingcow to get this over the line, including some testing on Mac?

Total explosion on the mac, viewer renders black. Nothing helpful sadly.

```Traceback (most recent call last):
  File "/Volumes/Data/dev/gaffer-pr-build/python/GafferUI/GLWidget.py", line 414, in drawBackground
    GL.glPopClientAttrib()
  File "/Volumes/Data/dev/gaffer-pr-build/python/OpenGL/error.py", line 232, in glCheckError
    baseOperation = baseOperation,
OpenGL.error.GLError: GLError(
	err = 1282,
	description = 'invalid operation',
	baseOperation = glPopClientAttrib,
	cArguments = ()```

I think we're going to have to check availability, and manage the state of the options accordingly.

themissingcow commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -45,6 +45,57 @@
 import GafferImage
 import GafferImageUI
 
+class DisplayTransformPlugValueWidget( GafferUI.PresetsPlugValueWidget ) :
```
We make these internal widgets protected `_DisplayTransformPlugValueWidget`

danieldresser-ie commented:
I have wasted way too much time on this, between the slow turn around time of testing the GL texture stuff on Mac, and stupidly forgetting that there's a prerequisite in this one line Cortex fix: https://github.com/ImageEngine/cortex/pull/1064

But hopefully, if you try running this locally with the cortex fix, you might see something now.

The test are currently on failing - on Linux just because of a new message warning that it's using terrible quality textures for image visualization on GPU, on Mac because of a still unknown GL error - but after trying to slowly investigate that, I think it may actually be caused by not having that Cortex fix ( though I'm not sure why this would affect Mac and Linux differently )

themissingcow commented:
> But hopefully, if you try running this locally with the cortex fix, you might see something now.

Nice work @danieldresser-ie this works a.ok on my mac now, and performance is _greatly_ improved for the Gamma/Exposure controls, even with the integrated GPU. An unscientific test with IPR tumbling definitely seem to improve interactivity too on my oversubscribed machine.

In case its any use:

```
DEBUG PRINT EXTENSIONS : GL_ARB_color_buffer_float GL_ARB_depth_buffer_float GL_ARB_depth_clamp GL_ARB_depth_texture GL_ARB_draw_buffers GL_ARB_draw_elements_base_vertex GL_ARB_draw_instanced GL_ARB_fragment_program GL_ARB_fragment_program_shadow GL_ARB_fragment_shader GL_ARB_framebuffer_object GL_ARB_framebuffer_sRGB GL_ARB_half_float_pixel GL_ARB_half_float_vertex GL_ARB_instanced_arrays GL_ARB_multisample GL_ARB_multitexture GL_ARB_occlusion_query GL_ARB_pixel_buffer_object GL_ARB_point_parameters GL_ARB_point_sprite GL_ARB_provoking_vertex GL_ARB_seamless_cube_map GL_ARB_shader_objects GL_ARB_shader_texture_lod GL_ARB_shading_language_100 GL_ARB_shadow GL_ARB_sync GL_ARB_texture_border_clamp GL_ARB_texture_compression GL_ARB_texture_compression_rgtc GL_ARB_texture_cube_map GL_ARB_texture_env_add GL_ARB_texture_env_combine GL_ARB_texture_env_crossbar GL_ARB_texture_env_dot3 GL_ARB_texture_float GL_ARB_texture_mirrored_repeat GL_ARB_texture_non_power_of_two GL_ARB_texture_rectangle GL_ARB_texture_rg GL_ARB_transpose_matrix GL_ARB_vertex_array_bgra GL_ARB_vertex_blend GL_ARB_vertex_buffer_object GL_ARB_vertex_program GL_ARB_vertex_shader GL_ARB_window_pos GL_EXT_abgr GL_EXT_bgra GL_EXT_blend_color GL_EXT_blend_equation_separate GL_EXT_blend_func_separate GL_EXT_blend_minmax GL_EXT_blend_subtract GL_EXT_clip_volume_hint GL_EXT_debug_label GL_EXT_debug_marker GL_EXT_draw_buffers2 GL_EXT_draw_range_elements GL_EXT_fog_coord GL_EXT_framebuffer_blit GL_EXT_framebuffer_multisample GL_EXT_framebuffer_multisample_blit_scaled GL_EXT_framebuffer_object GL_EXT_framebuffer_sRGB GL_EXT_geometry_shader4 GL_EXT_gpu_program_parameters GL_EXT_gpu_shader4 GL_EXT_multi_draw_arrays GL_EXT_packed_depth_stencil GL_EXT_packed_float GL_EXT_provoking_vertex GL_EXT_rescale_normal GL_EXT_secondary_color GL_EXT_separate_specular_color GL_EXT_shadow_funcs GL_EXT_stencil_two_side GL_EXT_stencil_wrap GL_EXT_texture_array GL_EXT_texture_compression_dxt1 GL_EXT_texture_compression_s3tc GL_EXT_texture_env_add GL_EXT_texture_filter_anisotropic GL_EXT_texture_integer GL_EXT_texture_lod_bias GL_EXT_texture_rectangle GL_EXT_texture_shared_exponent GL_EXT_texture_sRGB GL_EXT_texture_sRGB_decode GL_EXT_timer_query GL_EXT_transform_feedback GL_EXT_vertex_array_bgra GL_APPLE_aux_depth_stencil GL_APPLE_client_storage GL_APPLE_element_array GL_APPLE_fence GL_APPLE_float_pixels GL_APPLE_flush_buffer_range GL_APPLE_flush_render GL_APPLE_object_purgeable GL_APPLE_packed_pixels GL_APPLE_pixel_buffer GL_APPLE_rgb_422 GL_APPLE_row_bytes GL_APPLE_specular_vector GL_APPLE_texture_range GL_APPLE_transform_hint GL_APPLE_vertex_array_object GL_APPLE_vertex_array_range GL_APPLE_vertex_point_size GL_APPLE_vertex_program_evaluators GL_APPLE_ycbcr_422 GL_ATI_separate_stencil GL_ATI_texture_env_combine3 GL_ATI_texture_float GL_ATI_texture_mirror_once GL_IBM_rasterpos_clip GL_NV_blend_square GL_NV_conditional_render GL_NV_depth_clamp GL_NV_fog_distance GL_NV_light_max_exponent GL_NV_texgen_reflection GL_NV_texture_barrier GL_SGIS_generate_mipmap GL_SGIS_texture_edge_clamp GL_SGIS_texture_lod 
GOT ARB float
```

danieldresser-ie commented:
OK, hopefully this is finally good to go, though we won't know for sure until we get Cortex updated in the dependency package.

While switching to your proposed UI Tom, I wasn't sure where to put the Alt G shortcut for discoverability since it's no longer associated with one menu item, so I've currently just removed it.  Having the menu separated makes it a bit less horrid to try and use muscle memory to toggle back and forth anyway.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -519,6 +475,57 @@ def __setValue( self, value, *unused ) :
+
+	def _updateFromPlug( self ) :
+
+		with Gaffer.Context() :
```
Why not `with self.getContext()`?

danieldresser-ie replied:
Addressed in 9dfafc7

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -519,6 +475,57 @@ def __setValue( self, value, *unused ) :
+
+		with Gaffer.Context() :
+
+			self.__button.setImage( "lutGPU.png" if self.getPlug().getValue() else "lutCPU.png" )
```
When switching to GPU, I'm not seeing the image update immediately. It seems a redraw isn't requested until after I subsequently move the mouse. This is on Linux. Do you see the same?

danieldresser-ie replied:
This my main remaining concern at this point - I haven't been able to reproduce this no matter how much I make sure the cursor isn't moving, and I'm not sure what would be happening.

johnhaddon replied:
I've spent a bit longer digging into this, after rebasing on master to be sure there wasn't any interaction with recent button highlighting fixes. I've now convinced myself that it has nothing whatsoever to do with mouse movement, and it's just that the first redraw is a little slow when switching to GPU mode. VTune implies that the delay is just OCIO generating the LUT, so I don't think we should worry about this. Sorry for the bother.

What is odd is that there is a redraw reported after we update the icon and before the redraw that pays for the LUT. So I don't know why Qt hasn't updated the icon visibly at that point, but that's a rabbit hole we definitely should not go down.

johnhaddon commented on python/GafferImageUITest/ImageGadgetTest.py:
```diff
@@ -95,7 +97,14 @@ def testDestroyWhileProcessing( self ) :
+
+		if len( mh.messages ):
+			for m in mh.messages:
+				print( "IGNORING MESSAGE: " + m.context + " : " + m.message )
```
Don't clutter the test output - a successful test should be silent. Instead, let's assert that there are either no messages, or the single message we expect.

danieldresser-ie replied:
Addressed in a26b343

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -164,14 +113,25 @@ def _updateFromPlugs( self ) :
+			""",
+
+			"plugValueWidget:type", "GafferImageUI.ImageViewUI._LutGPUPlugValueWidget",
+			"toolbarLayout:index", 5,
```
Might it make more sense to put this after the displayTransform dropdown rather than before? So it reads "apply this exposure/gamma/clipping/transform using this method"?

danieldresser-ie replied:
Addressed by Tom.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -164,14 +113,25 @@ def _updateFromPlugs( self ) :
+		"displayTransformGPU" : [
+			"description",
+			"""
+			Applies a gamma correction to the image.
```
This description doesn't apply.

danieldresser-ie replied:
Addressed in 9dfafc7

themissingcow commented:
@danieldresser-ie. @johnhaddon and I were chatting about my rubbish icon this morning. We came up with this for a plan:

![image](https://user-images.githubusercontent.com/896779/87558395-f7134380-c6b0-11ea-9b3f-8bce49af85ae.png)

As in reality, it affects `exposure`, `gamma` _and_ `displayTransform`. As such, placement to the left of the menu is a little misleading. Available here: https://github.com/themissingcow/gaffer/commits/gpuOcioGraphics.

We thought it'd be good then to just disable that widget when the GPU isn't available with a suitable tool tip. Let me know if the icon doesn't look different enough with `setEnabled( False )`.

Whilst I was testing this, I noticed that in a `ShaderView` the scene menu bleeds into this group (always has). I took the liberty of adding this https://github.com/themissingcow/gaffer/commit/9e072f8e090439373fbd1752ab04b24efd701c5a (@johnhaddon may well have thoughts there), but it feels like we need _something_ like this to support these situations (see  https://github.com/themissingcow/gaffer/commit/dfb9bfd1133ec508795da498def3bcab060904fa).

So now we have:

![image](https://user-images.githubusercontent.com/896779/87558665-45c0dd80-c6b1-11ea-965f-e5c935a6396d.png)

johnhaddon commented:
> Whilst I was testing this, I noticed that in a ShaderView the scene menu bleeds into this group (always has). I took the liberty of adding this themissingcow@9e072f8 (@johnhaddon may well have thoughts there), but it feels like we need something like this to support these situations (see themissingcow@dfb9bfd).

The "divider" metadata is a bit of a strange one anyway, and the feedback we have is that folks would rather have a divider widget that they can drag round independently in the UIEditor than some metadata on a plug. So I'm reluctant to add yet another bit of metadata that isn't quite what we want. Can't we just register `"divider", True` for the plug on the ShaderView, but not on the ImageView?

themissingcow commented:
> > Whilst I was testing this, I noticed that in a ShaderView the scene menu bleeds into this group (always has). I took the liberty of adding this [themissingcow/gaffer@9e072f8](https://github.com/themissingcow/gaffer/commit/9e072f8) (@johnhaddon may well have thoughts there), but it feels like we need something like this to support these situations (see [themissingcow/gaffer@dfb9bfd](https://github.com/themissingcow/gaffer/commit/dfb9bfd)).
> 
> The "divider" metadata is a bit of a strange one anyway, and the feedback we have is that folks would rather have a divider widget that they can drag round independently in the UIEditor than some metadata on a plug. So I'm reluctant to add yet another bit of metadata that isn't quite what we want. Can't we just register `"divider", True` for the plug on the ShaderView, but not on the ImageView?

Thats a better idea! I forgot we could do that. Neat. @danieldresser-ie are you ok to do that when you cherry pick the bits you want from that branch? Cheers!

danieldresser-ie commented on src/GafferImageUI/ImageView.cpp:
```diff
@@ -557,10 +557,6 @@ bool ImageView::keyPress( const GafferUI::KeyEvent &event )
-	else if( event.key == "G" && event.modifiers == ModifiableEvent::Modifiers::Alt )
-	{
-		displayTransformGPUPlug()->setValue( !displayTransformGPUPlug()->getValue() );
-	}
```
I've reverted this.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -519,6 +475,57 @@ def __setValue( self, value, *unused ) :
+				"/" + name,
+				{
+					"command" : functools.partial( Gaffer.WeakMethod( self.__setValue ), value ),
+					"checkBox" : lutGPU == value
```
Addressed in 9dfafc7

danieldresser-ie commented on Changes.md:
```diff
@@ -5,6 +5,7 @@ Features
 --------
 
 - InteractiveRender : Added message log to the node's UI, displaying output from the last render (#3419).
+- ImageView : Support OCIO display transform and gamma/exposure on GPU for improved performance
```
Addressed in 380dfed

danieldresser-ie commented on python/GafferImageUITest/ImageGadgetTest.py:
```diff
@@ -95,7 +97,14 @@ def testDestroyWhileProcessing( self ) :
+
+		if len( mh.messages ):
+			for m in mh.messages:
+				print( "IGNORING MESSAGE: " + m.context + " : " + m.message )
```
Addressed in a26b343

danieldresser-ie commented on python/GafferUI/PresetsPlugValueWidget.py:
```diff
@@ -54,7 +54,7 @@ def __init__( self, plugs, **kw ) :
 		GafferUI.PlugValueWidget.__init__( self, cont, plugs, **kw )
 
-		self.__menuButton = GafferUI.MenuButton( "", menu = GafferUI.Menu( Gaffer.WeakMethod( self.__menuDefinition ) ) )
+		self.__menuButton = GafferUI.MenuButton( "", menu = GafferUI.Menu( Gaffer.WeakMethod( self._presetsMenuDefinition ) ) )
```
Reverted

danieldresser-ie commented:
OK, this should be pretty close to good to go, I think I've got the UI to a point which hopefully satisfies all the suggestions.  Main concern is John's delayed redraw thing, which I'm not sure how to tackle since I can't reproduce it.

Also not sure how I'm supposed to handle the merge conflict on graphics.svg?

johnhaddon commented:
> not sure how I'm supposed to handle the merge conflict on graphics.svg?

Sometimes I can just about manage to resolve the conflict by editing the text, but not in this case. Then I usually resort to using Inkscape to copy the new graphics out and in again. Since I had to rebase on master for testing anyway, I've done this in https://github.com/johnhaddon/gaffer/commit/f6a7070961a7c780358745ec5ef4408209861401. You can simply drop your commit and cherry-pick that one in its place. Please do take that exact commit, because I've also fixed a small issue with the bounding box for the graphics.

This all LGTM and I've resolved all my previous comments. I did notice though that the ShaderView still has the GPU icon and the divider in the wrong place though. Could you fix that at the same time as either addressing any further comments Tom has, or squashing down for merge? @themissingcow, could you go through your previous review and either resolve/approve or note what still needs doing?

johnhaddon commented:
> I did notice though that the ShaderView still has the GPU icon and the divider in the wrong place though.

Actually, I see that that is fine on your branch, but not on my rebased branch. I suspect that it has been messed up by the rejig of the pause/status icon for the SceneViewInspector, which we merged recently. So it still needs fixing, just you'll need to rebase to see the problem, and then fix.

danieldresser-ie commented:
OK, I've fixed the layout index and squashed, this should be good to go.

johnhaddon commented:
Thanks Daniel - I've pushed a rebase to smoosh a couple of commits together where they would otherwise have made bisecting hard (broken build). Will merge once CI is done.
